app/controllers/users/sessions_controller.rb

```diff
@@ -31,15 +31,21 @@ def new
 
     def create
       session[:sign_in_flow] = :sign_in
-      return process_locked_out_session if session_bad_password_count_max_exceeded?
-      return process_locked_out_user if current_user && user_locked_out?(current_user)
+
+      failure_reason = nil
+      failure_reason = process_locked_out_session if session_bad_password_count_max_exceeded?
+      failure_reason = process_locked_out_user if current_user && user_locked_out?(current_user)
+      failure_reason = process_failed_captcha if !valid_captcha_result?
+      return if failure_reason
 
       rate_limit_password_failure = true
+      failure_reason = :invalid_credentials
```

Comment:
Not sure why we set the failure reason as invalid credentials here before we authenticate and then set it as nil?

Reply:
`warden.authenticate!` will raise an exception if the authentication is invalid, so assigning the failure reason before will use that as the failure reason in the `ensure` block if the exception is raised. It's reset to `nil` afterward since that means the authentication was valid (no error raised).

Kinda confusing, I'll either refactor per @zachmargolis's comment, or at least add a code comment.

Reply:
I updated this in a49374ef31 which results in a much smaller diff and avoids juggling `failure_reason`.